src/sycltest/SYCLCore/ScopedContext.h

```diff
@@ -0,0 +1,233 @@
+      explicit ScopedContextAcquire(const ProductBase& data,
+                                    edm::WaitingTaskWithArenaHolder waitingTaskHolder,
+                                    ContextState& state)
+          : ScopedContextGetterBase(data), holderHelper_{std::move(waitingTaskHolder)}, contextState_{&state} {}
+
+      ~ScopedContextAcquire();
+
+      template <typename F>
+      void pushNextTask(F&& f) {
+        if (contextState_ == nullptr)
+          throwNoState();
+        holderHelper_.pushNextTask(std::forward<F>(f), contextState_);
+      }
+
+      void replaceWaitingTaskHolder(edm::WaitingTaskWithArenaHolder waitingTaskHolder) {
+        holderHelper_.replaceWaitingTaskHolder(std::move(waitingTaskHolder));
+      }
+
+    private:
+      void throwNoState();
+
+      impl::ScopedContextHolderHelper holderHelper_;
+      ContextState* contextState_ = nullptr;
+    };
+
+    /**
+     * The aim of this class is to do necessary per-event "initialization" in ExternalWork produce() or normal produce():
+     * - setting the current device
+     * - synchronizing between CUDA streams if necessary
+     * and enforce that those get done in a proper way in RAII fashion.
+     */
+    class ScopedContextProduce : public impl::ScopedContextGetterBase {
+    public:
+      /// Constructor to create a new CUDA stream (non-ExternalWork module)
+      explicit ScopedContextProduce(edm::StreamID streamID) : ScopedContextGetterBase(streamID) {}
+
+      /// Constructor to (possibly) re-use a CUDA stream (non-ExternalWork module)
+      explicit ScopedContextProduce(const ProductBase& data) : ScopedContextGetterBase(data) {}
+
+      /// Constructor to re-use the CUDA stream of acquire() (ExternalWork module)
+      explicit ScopedContextProduce(ContextState& state)
+          : ScopedContextGetterBase(state.releaseStream()) {}
+
+      /// Record the CUDA event, all asynchronous work must have been queued before the destructor
+      ~ScopedContextProduce();
+
+      template <typename T>
+      std::unique_ptr<Product<T>> wrap(T data) {
+        // make_unique doesn't work because of private constructor
+        return std::unique_ptr<Product<T>>(new Product<T>(stream(), event_, std::move(data)));
+      }
+
+      template <typename T, typename... Args>
+      auto emplace(edm::Event& iEvent, edm::EDPutTokenT<T> token, Args&&... args) {
+        return iEvent.emplace(token, stream(), event_, std::forward<Args>(args)...);
+      }
+
+    private:
+      friend class cudatest::TestScopedContext;
+
+      // This construcor is only meant for testing
+      explicit ScopedContextProduce(sycl::queue stream, sycl::event event)
+          : ScopedContextGetterBase(stream), event_{event} {}
+
+      // the barrier should be a no-op on an ordered queue, but is used to initialise the event on the data stream
+      sycl::event event_ = stream().submit_barrier();
```

Comment:
It does:

> `event()`:
> Constructs a host `event` that is immediately ready.  The `event` has no dependencies and no associated commands.   > Waiting on this `event` will return immediately.

Reply:
However that is associated to the wrong device and queue, so I'm not sure it's the best thing to use.

Reply:
That would actually be fine, because the `event_` is not used for synchronization before the destructor. Wait. That makes me realize that copying the event usage pattern from CUDA does not actually work here.

With CUDA we create the event in the `ScopedContextProduce` constructor, distribute a pointer to it to all the `Product<T>`s, and record the event in the destructor of `ScopedContextProduce`.

Here, an event is created in the `ScopedContextProduce` constructor, it is copied to all the `Product<T>`s, and a new event is assigned in the `ScopedContextProduce` destructor. But that event is now disconnected from the events of the `Product<T>`s.

Instead, we should do (e.g.) something like constructing a `std::shared_ptr<sycl::event>`in the constructor (the default constructor would be fine, IIUC), distributing the `shared_ptr` to the `Product<T>`s, and in the destructor assign the event to the de-referenced `shared_ptr`.

Written that, I would be fine proceeding with the current (broken) mode and fixing it in a subsequent PR (and I could volunteer to do it).